### PR TITLE
EASY-2430: reject javascript links

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes EASY-2430: reject javascript links in ddm

#### When applied it will
*
* 
* 

#### Related pull requests on github
https://github.com/DANS-KNAW/dans-bagit-profile/pull/6

### CODE REVIEW

File: src/test/scala/nl.knaw.dans.easy.validatebag/rules/metadata/MetadataRulesSpec.scala [hunk 1]
```diff
@@ -162,6 +163,24 @@ class MetadataRulesSpec extends TestSupportFixture with CanConnectFixture {
...
+      <ddm:replaces href="http://dans.knaw.nl">http://dans.knaw.nl</ddm:replaces>
+      <ddm:isRequiredBy scheme="URI">https://easy.dans.knaw.nl</ddm:isRequiredBy>
+   )) shouldBe a[Success[_]]
+  }
+
+  it should "report invalid link protocols" in {
+    ddmLinksHaveValidProtocol(bagWithExtraDcmi(
+      <ddm:conformsTo scheme="URL">javascript:alert('XSS')</ddm:conformsTo>
+      <ddm:conformsTo scheme="URN">ldap://localhost</ddm:conformsTo>
+      <ddm:conformsTo href="ftp://dans.knaw.nl/no/download/exists">ftp://dans.knaw.nl/no/download/exists</ddm:conformsTo>
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.validatebag/rules/metadata/MetadataRulesSpec.scala]: Maybe I should only validate href.

File: src/main/scala/nl.knaw.dans.easy.validatebag/rules/metadata/package.scala [hunk 2]
```diff
@@ -130,6 +130,32 @@ package object metadata extends DebugEnhancedLogging {
...
+        .filter(isLink)
+        .filterNot(validLinkProtocol)
+      if (invalidLinks.nonEmpty)
+        fail(s"Invalid link protocols: ${ invalidLinks.mkString(", ") }")
+    }
+  }
+
+  private def isLink(node: Node) = {
+    node.attributes.exists(
+      attribute => attribute.key == "scheme" || attribute.key == "href"
```
[the remark this answers is not in the document]
Comment (rvanheest) [reply src/main/scala/nl.knaw.dans.easy.validatebag/rules/metadata/package.scala]: and what about `<dcterms:license xsi:type="dcterms:URI">http://creativecommons.org/licenses/by-nc-sa/4.0/</dcterms:license>` or elements with `xsi:type="dcterms:URI"` in general?

**Comment by rvanheest** [comment]
Other question (@janvanmansum @lindareijnhoudt @jo-pol) what if `scheme="..."` tells us the `node.text` is something different than a URI/URL? (1) is that possible, (2) should we account for that?

File: src/test/scala/nl.knaw.dans.easy.validatebag/TestSupportFixture.scala [hunk 3]
```diff
@@ -57,12 +58,47 @@ trait TestSupportFixture extends FlatSpec with Matchers with Inside with BeforeA
     inside(result) {
       case Failure(e: RuleViolationDetailsException) =>
         e.getMessage should include(includedInErrorMsg)
-      case Failure(e) => fail(s"Not the expected type of exception: $e")
+      case Failure(e) => fail(s"Expecting [$includedInErrorMsg] got: $e")
```
[the remark this answers is not in the document]
Comment (jo-pol) [reply src/test/scala/nl.knaw.dans.easy.validatebag/TestSupportFixture.scala]: No, you get that only with shoulBe. As this test is in a fixture and and not in a spec it is harder to find the code without the expected value. Discovered by an abandoned variant for the new test.

File: src/test/scala/nl.knaw.dans.easy.validatebag/rules/metadata/MetadataRulesSpec.scala [hunk 4]
```diff
@@ -162,6 +163,24 @@ class MetadataRulesSpec extends TestSupportFixture with CanConnectFixture {
...
+    )) shouldBe Failure(RuleViolationDetailsException(
```
[the remark this answers is not in the document]
Comment (jo-pol) [reply src/test/scala/nl.knaw.dans.easy.validatebag/rules/metadata/MetadataRulesSpec.scala]: Started with those but see my reaction above. Those methods make trouble shooting harder. Perhaps a custom matcher that uses the argument for `bagWithExtraDcmi`. But then it is harder to mix valid an invalid input.

[the remark this answers is not in the document]
Comment (jo-pol) [reply src/test/scala/nl.knaw.dans.easy.validatebag/rules/metadata/MetadataRulesSpec.scala]: My approach doesn't repeat the BagIt validation. But that is a waste anyhow when testing variations in DDM.

**Comment by janvanmansum** [comment]
Closing for now. The requirements are being formulated in EASY-2452.

**Pull request closed without merging**